File: src/game/state/Harvesting.cpp
//
// Created by Per-Arne on 26.02.2017.
//

#include <iostream>
#include "Harvesting.h"
#include "../unit/unit.h"
#include "../util/Pathfinder.h"
#include "../player/Player.h"

void Harvesting::update(std::shared_ptr<Unit> unit)const{

    if (unit->harvestIterator == 0) {
        // Go to harvestable_tile;

        if(unit->distance(*unit->harvestTarget) > 1) {
            // Must walk
            // Find closest walkable tile to the target
            Tile* closestWalkable = Pathfinder::find_first_walkable_tile(unit->harvestTarget);

            // If distance is > 1, change harvest_target
            if(closestWalkable->distance(unit->harvestTarget) > 1) {
                Tile *closestHarvestable = Pathfinder::find_first_harvestable_tile(closestWalkable);
                if(!closestHarvestable){
                    unit->transitionState(unit->stateManager.idleState);
                    return;
                }
                unit->harvestTarget = closestHarvestable;


            }

            unit->walkingGoal = closestWalkable;
            unit->transitionState(unit->stateManager.walkingState);
            unit->enqueueState(unit->stateManager.harvestingState);
        }

        unit->harvestIterator = 1;




    } else if (unit->harvestIterator == 1) {
        // Start harvesting

        unit->harvestTimer += 1;
        if(unit->harvestTimer >= unit->harvestInterval) {

            if(unit->lumberCarry + unit->goldCarry + unit->oilCarry >= unit->carryCapacity) {
                // Return to base
                unit->harvestIterator = 2;
                return;
            }

            unit->lumberCarry += unit->harvestTarget->lumberYield;
            unit->goldCarry += unit->harvestTarget->goldYield;
            unit->oilCarry += unit->harvestTarget->oilYield;
            unit->player_.statGoldGather += unit->harvestTarget->goldYield;
            unit->player_.statLumberGather += unit->harvestTarget->lumberYield;
            unit->player_.statOilGather += unit->harvestTarget->oilYield;
            unit->harvestTarget->resources -= unit->harvestTarget->lumberYield + unit->harvestTarget->goldYield + unit->harvestTarget->oilYield;

            // No more resources // TODO constant parameter Config
            if(unit->harvestTarget->resources < 1) {
                unit->harvestTarget->setDepleted();

                // Find new harvestable
                Tile *closestHarvestable = Pathfinder::find_first_harvestable_tile(unit->harvestTarget);
                if(!closestHarvestable){
                    unit->transitionState(unit->stateManager.idleState);
                    return;
                }


                unit->harvestTarget = closestHarvestable;
            }

            unit->harvestTimer = 0;



        }


    } else if (unit->harvestIterator == 2) {
        // Recall (Walking)
        std::shared_ptr<Unit> closestBase = unit->closestRecallBuilding();

        // No base to recall to
        if (!closestBase) {
            unit->transitionState();
            return;;
        }


        if(unit->distance(closestBase) > 1) {
            // Must walk
            unit->walkingGoal = closestBase->tile;
            unit->transitionState(unit->stateManager.walkingState);
            unit->enqueueState(unit->stateManager.harvestingState);
            return;
        } else {
            // Can deliver
            unit->player_.addGold(unit->goldCarry);
            unit->player_.addLumber(unit->lumberCarry);
            unit->player_.addOil(unit->oilCarry);

            unit->goldCarry = 0;
            unit->lumberCarry = 0;
            unit->oilCarry = 0;

            if(!Config::getInstance().getMechanicHarvestReplay())
            {
                unit->transitionState(unit->stateManager.harvestingState);
            }else{
                unit->transitionState();
            }
        }

    }
}

void Harvesting::end(std::shared_ptr<Unit> unit)const{

}

void Harvesting::init(std::shared_ptr<Unit> unit)const{
    unit->harvestTimer = 0;
    unit->harvestIterator = 0;
}

File: python/game/state/Harvesting.py
from game import Config
from game import const
from game.state.GenericState import GenericState




class Harvesting(GenericState):
    id = "Harvesting"
    type = const.State.ID_Harvesting


    # Harvesting
    harvest_interval = .5 * Config.FRAME_MULTIPLIER
    harvest_timer = 0
    harvest_iterator = 1

    RETURN = 0
    HARVEST = 1

    harvest_target = None

    def __init__(self, unit):
        super().__init__(unit)
        self.data = unit.d['s'][Harvesting.type]

    def init(self):
        self.data['harvest_timer'] = 0

    def stage_goto_harvest(self):
        tiles = self.game.AdjacentMap.adjacent_walkable(self.game, *self.data['harvest_target'], 1)

        if not tiles:
            return
        tile = self.unit.shortest_distance(tiles)

        if not self.has_next_state(const.State.ID_Harvesting):
            self.add_next(const.State.ID_Harvesting, {
                'harvest_target': self.data['harvest_target']
            }, 1)
            self.unit.move(*tile)

    def stage_harvest(self):
        if self.data['harvest_timer'] >= self.harvest_interval:

            tile_data = self.Map.get_tile(*self.data['harvest_target'])

            if self.unit.d['inventory_lumber'] + \
                    self.unit.d['inventory_gold'] + \
                    self.unit.d['inventory_oil'] >= self.unit.inventory_space:
                # Inventory is full, return to base
                recall_building = self.unit.closest_recall_building()
                if not recall_building:
                    # Transition to next state since nothing can be done
                    self.transition()
                    return
                else:
                    self.data['harvest_iterator'] = Harvesting.RETURN
                    return

            self.unit.d['inventory_gold'] += tile_data['gold_yield']
            self.unit.d['inventory_lumber'] += tile_data['lumber_yield']
            self.unit.d['inventory_oil'] += tile_data['oil_yield']

            # Todo, adding statistics here becaue its thought to be better for RL????
            self.player.statistics['gold_count'] += tile_data['gold_yield']
            self.player.statistics['lumber_count'] += tile_data['lumber_yield']
            self.player.statistics['oil_count'] += tile_data['oil_yield']

            self.data['harvest_timer'] = 0

    def stage_recall(self, recall_building):
        # Must walk
        tiles = self.game.AdjacentMap.adjacent_walkable(self.game, recall_building.d['x'] + recall_building.dimension,
                                                       recall_building.d['y'] + recall_building.dimension,
                                                       recall_building.dimension + 1)

        if tiles:
            # Wait for free spot to deliver
            tile = self.unit.shortest_distance(tiles)
            if not self.has_next_state(const.State.ID_Harvesting):
                self.unit.move(*tile)
                self.add_next(const.State.ID_Harvesting, {
                    'harvest_iterator': Harvesting.RETURN,
                }, 1)


    def stage_deliver(self):
        self.unit.player.gold += self.unit.d['inventory_gold']
        self.unit.player.lumber += self.unit.d['inventory_lumber']
        self.unit.player.oil += self.unit.d['inventory_oil']
        self.unit.d['inventory_gold'] = 0
        self.unit.d['inventory_lumber'] = 0
        self.unit.d['inventory_oil'] = 0
        self.data['harvest_iterator'] = Harvesting.HARVEST

    def update(self, tick):
        self.data['harvest_timer'] += tick

        if self.data['harvest_iterator'] == Harvesting.HARVEST:

            distance = self.unit.distance(*self.data['harvest_target'])
            if distance > 1:
                """
                # Unit cannot reach the harvest tile, It therefore must trigger a Walking state to get to the tile
                """
                self.stage_goto_harvest()
            else:
                """
                # Unit can reach the harvest tile. Start harvesting
                """
                self.stage_harvest()

        elif self.data['harvest_iterator'] == Harvesting.RETURN:
            """
            # Find closest recall building (Typically a town-center or lumber mill
            """
            recall_building = self.unit.closest_recall_building()

            if recall_building is None:
                self.transition()
                return

            distance = self.unit.distance(
                recall_building.d['x'],
                recall_building.d['y'],
                recall_building.dimension
            )

            if distance > 1:
                """
                # Unit cannot reach the recall building, This means that the unit must move towards it
                # Triggers Walking state
                """
                self.stage_recall(recall_building)
            else:
                """
                # Unit are ready for delivering goods to the selected recall building
                # When delivery is done, trigger new harvesting round
                """
                self.stage_deliver()

                if Config.MECHANICS_HARVEST_FOREVER:
                    self.unit.harvest(*self.data['harvest_target'])
                else:
                    self.transition()
